tsfile/src/main/java/org/apache/tsfile/file/metadata/StringArrayDeviceID.java

```diff
@@ -0,0 +1,271 @@
+    for (String segment : segments) {
+      cnt += ReadWriteIOUtils.writeVar(segment, byteBuffer);
+    }
+    return cnt;
+  }
+
+  @Override
+  public int serialize(OutputStream outputStream) throws IOException {
+    int cnt = 0;
+    cnt += ReadWriteForEncodingUtils.writeUnsignedVarInt(segments.length, outputStream);
+    for (String segment : segments) {
+      cnt += ReadWriteIOUtils.writeVar(segment, outputStream);
+    }
+    return cnt;
+  }
+
+  public static StringArrayDeviceID deserialize(ByteBuffer byteBuffer) {
+    final int cnt = ReadWriteForEncodingUtils.readUnsignedVarInt(byteBuffer);
+    if (cnt == 0) {
+      return new StringArrayDeviceID(new String[] {""});
+    }
+
+    String[] segments = new String[cnt];
+    for (int i = 0; i < cnt; i++) {
+      segments[i] = ReadWriteIOUtils.readVarIntString(byteBuffer);
+    }
+    return new StringArrayDeviceID(segments);
+  }
+
+  public static StringArrayDeviceID deserialize(InputStream stream) throws IOException {
+    final int cnt = ReadWriteForEncodingUtils.readUnsignedVarInt(stream);
+    if (cnt == 0) {
+      return new StringArrayDeviceID(new String[] {""});
+    }
+
+    String[] segments = new String[cnt];
+    for (int i = 0; i < cnt; i++) {
+      segments[i] = ReadWriteIOUtils.readVarIntString(stream);
+    }
+    return new StringArrayDeviceID(segments);
+  }
+
+  @Override
+  public byte[] getBytes() {
+    ByteArrayOutputStream publicBAOS = new ByteArrayOutputStream(256);
+    for (String segment : segments) {
+      try {
+        publicBAOS.write(segment.getBytes(TSFileConfig.STRING_CHARSET));
+      } catch (IOException e) {
+        throw new TsFileRuntimeException(e);
+      }
+    }
+    return publicBAOS.toByteArray();
+  }
+
+  @Override
+  public boolean isEmpty() {
+    return segments == null || segments.length == 0;
+  }
+
+  @Override
+  public String getTableName() {
+    return segments[0];
+  }
+
+  @Override
+  public int segmentNum() {
+    return segments.length;
+  }
+
+  @Override
+  public String segment(int i) {
+    return segments[i];
+  }
+
+  @Override
+  public int compareTo(IDeviceID o) {
+    int thisSegmentNum = segmentNum();
+    int otherSegmentNum = o.segmentNum();
+    for (int i = 0; i < thisSegmentNum; i++) {
+      if (i >= otherSegmentNum) {
+        // the other ID is a prefix of this one
+        return 1;
+      }
+      final int comp =
+          Objects.compare(this.segment(i), ((String) o.segment(i)), WriteUtils::compareStrings);
+      if (comp != 0) {
+        // the partial comparison has a result
+        return comp;
+      }
+    }
+
+    if (thisSegmentNum < otherSegmentNum) {
+      // this ID is a prefix of the other one
+      return -1;
+    }
+
+    // two ID equal
+    return 0;
+  }
+
+  @Override
+  public long ramBytesUsed() {
+    return INSTANCE_SIZE + RamUsageEstimator.sizeOf(segments);
+  }
+
+  @Override
+  public int serializedSize() {
```

Comment:
while serializing string, we use varInt to represent string length and string array count, may only occupy just one byte.